### PR TITLE
Fixing bugs around scoping and rewriters

### PR DESCRIPTION
**Pull request by davidegrohmann**
This is a major overhaul of how we handle identifiers, aliased and
scoping.  The rewriters responsible for this work were not cooperating
well.  This commit fixes that by improving scoping rules, improving
some of the rewriter implementations and introducing extra tests.

### CODE REVIEW

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/SemanticState.scala [hunk 1]
```diff
@@ -19,11 +19,22 @@
...
 
+// A symbol use represents the occurrence of a symbol at a position
+case class SymbolUse(name: String, position: InputPosition) {
+  override def toString = s"SymbolUse($nameWithPosition)"
+
+  def nameWithPosition = s"$name@${position.offset}"
+}
+
+
+// MATCH n, x RETURN n.prop as x, x.bar as foo
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/SemanticState.scala]: let's clean up this...

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/Namespacer.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    val ambiguousNames = shadowedNames(scopeTree)
+    val identifierDefinitions: Map[SymbolUse, SymbolUse] = scopeTree.allIdentifierDefinitions
+    val protectedIdentifiers = returnAliases(statement)
+    val renamings = identifierRenamings(statement, identifierDefinitions, ambiguousNames, protectedIdentifiers)
+    Namespacer(renamings)
+  }
+
+  private def shadowedNames(scopeTree: Scope): Set[String] = {
+    val cleanScopeTree =
+      scopeTree//.copy(children = scopeTree.children.dropRight(1)) // HACK! This is the last RETURN clause, and should not be used to find shadowed identifiers
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/Namespacer.scala]: let's remove this comment

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/tracing/rewriters/RewriterTaskProcessor.scala [hunk 3]
```diff
@@ -50,7 +50,8 @@ trait RewriterTaskProcessor extends (RewriterTask => Rewriter) {
       for (problem <- failure.problems)
         builder ++= s"Condition '$name' violated. $problem"
     }
-    builder.toString()
+    val result = builder.toString()
+    result
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/tracing/rewriters/RewriterTaskProcessor.scala]: this is not needed

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/NamespacerTest.scala [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    //    |RETURN latestLike.likeTime AS likeTime
+    //    |ORDER BY likeTime
+    //    | """.stripMargin ->
+    //  """MATCH (person:Person       )<-               -(message)<-[like]-(liker:Person)
+    //    |WITH                 like.creationDate AS `  likeTime@137`, person AS person
+    //    |ORDER BY `  likeTime@137`         , message.id
+    //    |WITH        head(collect({              likeTime: `  likeTime@137`})) AS latestLike, person AS person
+    //    |WITH latestLike.likeTime AS `  likeTime@327`
+    //    |ORDER BY `  likeTime@327`
+    //    |RETURN `  likeTime@327` AS likeTime""".stripMargin
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/NamespacerTest.scala]: commented tests...

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor2_2.scala [hunk 5]
```diff
@@ -116,7 +116,7 @@ object exceptionHandlerFor2_2 extends MapToPublicExceptions[CypherException] {
 }
 
 trait CompatibilityFor2_2 {
-  import helpers._
+  import org.neo4j.cypher.internal.compatibility.helpers._
```
Comment (davidegrohmann) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor2_2.scala]: not needed

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/normalizeReturnClauses.scala [hunk 6]
```diff
@@ -56,18 +56,31 @@ case object normalizeReturnClauses extends Rewriter {
       )
 
     case clause @ Return(distinct, ri, orderBy, skip, limit) =>
-      val (aliasProjection, finalProjection) = ri.items.map(i => {
-        val newPosition = i.expression.position.copy(offset = i.expression.position.offset + 1)
-        if (i.alias.isDefined) {
-          (i, AliasedReturnItem(i.alias.get, i.alias.get.copy()(newPosition))(i.position))
-        } else {
+      var rewrites = Map[Expression, Expression]()
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/normalizeReturnClauses.scala]: Perhaps we can remove the var by pushing it in the unzip and return a triplet

**Comment by systay** [comment]
Adressed comments and pushed

**Pull request closed without merging**